system-addon/content-src/components/TopSites/TopSiteForm.jsx

```diff
@@ -12,13 +12,36 @@ export class TopSiteForm extends React.PureComponent {
     this.state = {
       label: site ? (site.label || site.hostname) : "",
       url: site ? site.url : "",
-      validationError: false
+      validationError: false,
+      customScreenshotUrl: site ? site.customScreenshotURL : "",
+      showCustomScreenshotForm: site ? site.customScreenshotURL : false,
+      screenshotRequestFailed: false,
+      screenshotPreview: null,
+      pendingScreenshotUpdate: false
     };
+    this.onClearScreenshotInput = this.onClearScreenshotInput.bind(this);
     this.onLabelChange = this.onLabelChange.bind(this);
     this.onUrlChange = this.onUrlChange.bind(this);
     this.onCancelButtonClick = this.onCancelButtonClick.bind(this);
     this.onClearUrlClick = this.onClearUrlClick.bind(this);
     this.onDoneButtonClick = this.onDoneButtonClick.bind(this);
+    this.onCustomScreenshotUrlChange = this.onCustomScreenshotUrlChange.bind(this);
+    this.onPreviewButtonClick = this.onPreviewButtonClick.bind(this);
+    this.onEnableScreenshotUrlForm = this.onEnableScreenshotUrlForm.bind(this);
+    this.validateUrl = this.validateUrl.bind(this);
+  }
+
+  componentWillReceiveProps(nextProps) {
+    // Check the response URL against the current state of the form
+    const pendingRequest = this.state.customScreenshotUrl && this.state.pendingScreenshotUpdate &&
+      nextProps.customScreenshotURL === this.cleanUrl(this.state.customScreenshotUrl);
+    if (pendingRequest || nextProps.screenshotRequestFailed) {
+      this.setState({
+        pendingScreenshotUpdate: false,
+        screenshotRequestFailed: nextProps.screenshotRequestFailed,
+        screenshotPreview: nextProps.screenshotPreview
```

Comment:
I've mentioned that having a copy of a prop as a state doesn't seem right and makes it hard to reason about when to use the value from state vs from prop especially when they're named exactly the same. It looks like the reason why it's in the state is so that it can be cleared from `onClearScreenshotInput` to get out of preview mode. Why couldn't `onClear` trigger something that makes the prop no longer be set, i.e., clear the local redux state.

Reply:
I'm no longer copying the props in the state and added a new action creator to modify the local reducer.
Removed `pendingScreenshotUpdate` and replaced it with `requestScreenshotURL` which is `true` at the same time + I can use it to validate the correct request is received.